src/hotspot/share/classfile/classLoader.cpp

```diff
@@ -826,14 +827,27 @@ bool ClassLoader::contains_append_entry(const char* name) {
   return false;
 }
 
+bool ClassPathEntry::cas_set_next(ClassPathEntry* next) {
+  return Atomic::replace_if_null(&_next, next);
+}
+
+// The boot append entries are added and read lock free.
 void ClassLoader::add_to_boot_append_entries(ClassPathEntry *new_entry) {
   if (new_entry != NULL) {
-    if (_last_append_entry == NULL) {
-      assert(_first_append_entry == NULL, "boot loader's append class path entry list not empty");
-      _first_append_entry = _last_append_entry = new_entry;
-    } else {
-      _last_append_entry->set_next(new_entry);
-      _last_append_entry = new_entry;
+    for (;;) {
+      ClassPathEntry* last_entry = Atomic::load(&_last_append_entry);
+      if (last_entry == NULL) {
+        if (Atomic::replace_if_null(&_last_append_entry, new_entry)) {
+          assert(first_append_entry() == NULL, "boot loader's append class path entry list not empty");
+          Atomic::store(&_first_append_entry_list, new_entry);
+          return;
```

Comment:
Not sure if it is a problem. There is a chance that an entry is added, e.g. _last_append_entry = new_entry, but _first_append_entry_list == NULL. If contains_append_entry() queries on this new entry, it may return false.

Reply:
That race also exists in the current code, since contains_append_entry() never acquired the same lock.  Even if it did, there's a race when one thread adds an entry and the other reads the list.

Reply:
void InstanceKlass::log_to_classlist(const ClassFileStream* stream) const {
  if (ClassListWriter::is_enabled()) {
  ...
        if (class_loader == NULL && ClassLoader::contains_append_entry(stream->source())) {
          // .. but don't skip the boot classes that are loaded from -Xbootclasspath/a
          // as they can be loaded from the archive during runtime.

in logging loaded classes from the NULL class loader, this is where contains_append_entry() races with a JVMTI thread concurrently adding a class path entry, but this is benign because it's logging a class that's already loaded, so it couldn't have loaded it from the partially added classpath.